src/genesis.rs

```diff
@@ -0,0 +1,306 @@
+        })
+    }
+
+    fn run(&mut self) -> impl futures::Future<Output = Result<()>> + Send {
+        self.start()
+    }
+}
+
+impl Genesis {
+    pub async fn start(&mut self) -> Result<(), Error> {
+        if !self
+            .config
+            .consensus
+            .genesis_stakers
+            .contains_key(&self.config.id)
+        {
+            return Ok(());
+        }
+
+        info!("🌱 Starting genesis block creation {:?}", self.config);
+        if self.config.id == self.config.consensus.genesis_leader {
+            info!(
+                "👑 Setting ourselves {}({}) as leader for genesis",
+                &self.config.id,
+                self.crypto.validator_pubkey()
+            );
+            self.first_round_leader = Some(self.crypto.validator_pubkey().clone());
+        }
+        // We will start from the genesis block.
+        self.peer_pubkey.insert(
+            self.config.id.clone(),
+            self.crypto.validator_pubkey().clone(),
+        );
+
+        // Wait until we've connected with all other genesis peers.
+        if self.config.id != "single-node" {
+            handle_messages! {
+                on_bus self.bus,
+                listen<PeerEvent> msg => {
+                    match msg {
+                        PeerEvent::NewPeer { name, pubkey } => {
+                            info!("🌱 New peer {} added to genesis: {}", &name, &pubkey);
+                            self.peer_pubkey
+                                .insert(name.clone(), pubkey.clone());
+
+                            if name == self.config.consensus.genesis_leader {
+                                info!("👑 Setting {}({}) as leader for genesis", &name, &pubkey);
+                                self.first_round_leader = Some(pubkey);
+                            }
+
+                            // Once we know everyone in the initial quorum, craft & process the genesis block.
+                            if self.peer_pubkey.len()
+                                == self.config.consensus.genesis_stakers.len() {
+                                break
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // At this point, we can setup the genesis block.
+        _ = self.bus.send(GenesisEvent::GenesisBlock {
+            initial_validators: self.peer_pubkey.values().cloned().collect(),
+            stake_txs: self
+                .peer_pubkey
+                .iter()
+                .map(|(k, v)| {
+                    Transaction::wrap(TransactionData::Stake(Staker {
+                        pubkey: v.clone(),
+                        stake: Stake {
+                            amount: *self.config.consensus.genesis_stakers.get(k).unwrap_or(&100),
+                        },
+                    }))
+                })
+                .collect(),
+        });
+
+        _ = self.bus.send(GenesisEvent::Ready {
```

Comment:
Mmh tu perds le fait qu'on attende que DA ait process ce genesis block là non (le code L1356-1370 de consensus.rs) ? Aucune garantie que le consensus va pas se lancer avant.

Reply:
ouep, mais vu qu'on fait le delay avant de start un slot ça passe, le traitement à faire pour ce bloc sera toujours assez rapide (sauf à mettre un delay bcp trop petit)

j'ai surtout pas tellement trouvé comment mieux faire pour la première itération (sans avoir une dépendance cachée et un comportement non explicite)

Reply:
Tu peux pas refaire ce que j'avais fait, à savoir listen le message `ProcessedBlock` venant de DA?

Reply:
en fait je suis en train de voir si j'arrive pas à remove l'event Ready, car dans le cas d'un node qui join un consensus en cours, il n'a pas de genesis à faire